### ISSUE TITLE
FBSDKApplicationDelegate in FacebookCore?

### ISSUE THREAD

**Issue by spikerola** [facebook/facebook-sdk-swift#35]
After the login and the permission agreement the web view is blank and the only button is "Done" which close and cancel the login. I followed the [developer guide](https://developers.facebook.com/docs/swift).

[see update 1](#update1)
[see last update (2)](#update2)

_Found_: Application delegate haven't the correct "openURL" methods..
**The guide is incorrect when they don't say you have to write those two rows..
But what is the native library "AppDelegate"?**

```swift
func application(application: UIApplication, openURL url: NSURL, sourceApplication: String?, annotation: AnyObject) -> Bool {
    return FBSDKApplicationDelegate.sharedInstance().application(application, openURL: url, sourceApplication: sourceApplication, annotation: annotation)
}

func application(app: UIApplication, openURL url: NSURL, options: [String : AnyObject]) -> Bool {
    return FBSDKApplicationDelegate.sharedInstance().application(app, openURL: url, sourceApplication: options[UIApplicationOpenURLOptionsSourceApplicationKey] as! String, annotation: options[UIApplicationOpenURLOptionsAnnotationKey])
}
```

[white web screen here](http://i.imgur.com/7OkCxIx.png)

<a name="update1"></a>

### Update 1

The first problem (White screen) was solved by adding the application delegate method (**Not mentioned in the [guide](https://developers.facebook.com/docs/swift)**)
The **second problem** is that with FBSDKApplicationDelegate works.. but with `ApplicationDelegate` (or if used into the ios application manager `FacebookCore.ApplicationDelegate`) the application crash because `UIApplicationOpenURLOptionsAnnotation` is nil.

<a name="update2"></a>

### Update 2 and recap

The problem of the **white web screen** after accepting the permission agreement was solved adding the facebook app delegate.
The second problem of the **applicationDelegate method** has been solved by replacing the `FBSDKApplicationDelegate` in functions
```swift
func application(application: UIApplication, openURL url: NSURL, sourceApplication: String?, annotation: AnyObject) -> Bool {
    return FBSDKApplicationDelegate.sharedInstance().application(application, openURL: url, sourceApplication: sourceApplication, annotation: annotation)
}

func application(app: UIApplication, openURL url: NSURL, options: [String : AnyObject]) -> Bool {
    return FBSDKApplicationDelegate.sharedInstance().application(app, openURL: url, sourceApplication: options[UIApplicationOpenURLOptionsSourceApplicationKey] as! String, annotation: options[UIApplicationOpenURLOptionsAnnotationKey])
}
```
with `FacebookCore.ApplicationDelegate` (that I hope it will be named _FBApplicationDelegate_) as it follow
```
func application(app: UIApplication, openURL url: NSURL, options: [String : AnyObject]) -> Bool {
    return FacebookCore.ApplicationDelegate.shared.application(app, openURL: url, sourceApplication: options[UIApplicationOpenURLOptionsSourceApplicationKey] as? String, annotation: options[UIApplicationOpenURLOptionsAnnotationKey] ?? [])
}
    
func application(application: UIApplication, openURL url: NSURL, sourceApplication: String?, annotation: AnyObject) -> Bool {
    return FacebookCore.ApplicationDelegate.shared.application(application, openURL: url, sourceApplication: sourceApplication, annotation: annotation)
}
```

> Note the `?? []`: 
> DO NOT USE `!`

Thanks to @nlutsenko 👍

**Comment by spikerola** [comment]
Found: Application delegate haven't the correct "openURL" methods

**Comment by nlutsenko** [comment]
The native Swift class to get the same functionality is `SDKApplicationDelegate` (without `FB` prefix).

**Comment by spikerola** [comment]
Gotcha

**Comment by spikerola** [comment]
No wait

**Comment by nlutsenko** [comment]
Swift is using the type system and inferred module semantics.
Better explained with an example:
In case if you don't have `ApplicationDelegate` in your Swift module - it will recognize `FacebookCore`'s one and give you an ability to use it.
If you have `ApplicationDelegate` type in your module - then you want to use `FacebookCore.ApplicationDelegate`, as that has explicit module reference.

Hope it makes sense now ;)
Looks like we might want to rename the methods there, though as we are likely to class with a generic `ApplicationDelegate`. Am going to send a patch in a few minutes.

**Comment by nlutsenko** [comment]
Also, as a hint - the call looks like this:
```swift
ApplicationDelegate.shared.application(application, didFinishLaunchingWithOptions: launchOptions)
```
and
```swift
ApplicationDelegate.shared.application(application,
                                                  openURL: url,
                                                  sourceApplication: sourceApplication,
                                                  annotation: annotation)
```

**Comment by spikerola** [comment]
I tried
```swift
return FacebookCore.ApplicationDelegate.shared.application(app, openURL: url, sourceApplication: options[UIApplicationOpenURLOptionsSourceApplicationKey] as? String, annotation: options[UIApplicationOpenURLOptionsAnnotationKey]!)
```
but `options[UIApplicationOpenURLOptionsAnnotationKey]` is nil...
and when I do this with 
```swift
return FBSDKApplicationDelegate.sharedInstance().application(app, openURL: url, sourceApplication: options[UIApplicationOpenURLOptionsSourceApplicationKey] as! String, annotation: options[UIApplicationOpenURLOptionsAnnotationKey])
```
works..

**Comment by nlutsenko** [comment]
Try the following:
```swift
return FacebookCore.ApplicationDelegate.shared.application(app, openURL: url, sourceApplication: options[UIApplicationOpenURLOptionsSourceApplicationKey] as? String, annotation: options[UIApplicationOpenURLOptionsAnnotationKey] ?? [])
```

Going to provide an implementation for a new method in a brief moment, since in case you are target iOS 9.0+ - the old one is deprecated.

**Comment by nlutsenko** [comment]
Let me know if the above method works for you with Swift SDK.

**Comment by spikerola** [comment]
Yeah it works! 😃 
![simulator screen shot 03 ago 2016 23 06 22](https://cloud.githubusercontent.com/assets/15785841/17381962/ebb05b5c-59ce-11e6-9158-1bf6ae09aca3.png)

**Comment by RajChanchal** [comment]
This works for me:
func application(_ app: UIApplication, open url: URL, options: [UIApplicationOpenURLOptionsKey : Any] = [:]) -> Bool {
        return FBSDKApplicationDelegate.sharedInstance().application(app, open: url, sourceApplication: options[UIApplicationOpenURLOptionsKey.sourceApplication] as! String, annotation: options[UIApplicationOpenURLOptionsKey.annotation])
    }

**Comment by AquilaSagitta** [comment]
Adding this to appDelegate fixed this issue for me

```
func application(_ app: UIApplication, open url: URL, options: [UIApplicationOpenURLOptionsKey : Any] = [:]) -> Bool {
        return FBSDKApplicationDelegate.sharedInstance().application(app, open: url as URL!, sourceApplication: options[UIApplicationOpenURLOptionsKey.sourceApplication] as! String, annotation: options[UIApplicationOpenURLOptionsKey.annotation])
    }
```

**Comment by borisyurkevich** [comment]
I got deprecated warning from Facebook SDK and  this fixed it.  🤔

    import FacebookCore
    import FacebookLogin
  
    public func application(_ app: UIApplication, open url: URL, options: [UIApplicationOpenURLOptionsKey : Any] = [:]) -> Bool {
    
        let isFacebookCanOpen = SDKApplicationDelegate.shared.application(app, open: url as URL!, options: options)
        return isFacebookCanOpen
    }

Tested on FBSDKCoreKit 4.20.0  
FBSDKLoginKit 4.20.0  
FacebookCore (0.2.0)  
Using FacebookLogin (0.2.0)

**Comment by spikerola** [comment]
@RajChanchal @AquilaSagitta @borisyurkevich 
Because now it's updated...

**Comment by savethedave** [comment]
@nlutsenko Could you clarify the latest AppDelegate functions for iOS 10? Thx.

**Comment by RajChanchal** [comment]
@savethedave following works fine in iOS 10:
```
func application(_ app: UIApplication, open url: URL, options: [UIApplicationOpenURLOptionsKey : Any] = [:]) -> Bool{
}
```

**Comment by savethedave** [comment]
Thx @RajChanchal you are correct. 

FWIW, I was following all guides and forums posts to the letter, but still having trouble. Then I reinstalled the SDK using Cocoapods instead of manual compiling, and it all worked fine. To be fair to FB, they do recommend using Cocoapods clearly in their docs, but other parts of the docs are obviously lacking, and probably wasting **thousands of developer hours monthly** - an unfortunate exception to the generally good docs FB has had in recent years.

For reference to others having trouble, here's the relevant parts of my AppDelegate file that are working for me as of this date (iOS 10, SDK 0.2.0):



    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        SDKApplicationDelegate.shared.application(application, didFinishLaunchingWithOptions: launchOptions)
        return true
    }

    func application(_ app: UIApplication, open url: URL, options: [UIApplicationOpenURLOptionsKey : Any]) -> Bool {
        let appId = SDKSettings.appId
        if url.scheme != nil && url.scheme!.hasPrefix("fb\(appId)") && url.host == "authorize" { // Facebook
            return SDKApplicationDelegate.shared.application(app, open: url, options: options)
        }
        return false
    }

**Comment by metisdev** [comment]
When 

    FacebookCore.SDKApplicationDelegate.shared.application(app, didFinishLaunchingWithOptions: launchOptions)


is called, do we also need to call :

    FacebookCore.AppEventsLogger.activate()

?

**Comment by spikerola** [comment]
Should I remember y'all that the issue is closed and it's related to an older version?

**Comment by hk0i** [comment]
I'm still getting this issue with FacebookCore 0.7.0

```swift
  func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
    if #available(iOS 9.0, *) {
      return SDKApplicationDelegate.shared.application(app, open: url, options: options)
    }

    return false
  }
```

and if I rename it tells me

> Module `FacebookCore` has no member named `ApplicationDelegate`

![FacebookCore has no member named ApplicationDelegate](https://user-images.githubusercontent.com/755669/59651778-0fc4c900-9140-11e9-86b1-8deffde9260a.png)

What is the current name? According to the docs it's `ApplicationDelegate`, but compiler seems to disagree.

**Comment by hk0i** [comment]
Fixed it by changing my import from `import FacebookCore` to `import FBSDKCoreKit` (the ObjC library) then you can use `ApplicationDelegate` without namespace issues.
